### PR TITLE
Added support for S3 storage classes

### PR DESCRIPTION
**Pull request by NadavTasher**
#### Summary
Add support for specifying the storage class to use for non-AWS S3-compatible storage solutions.

Some S3-compatible storage solutions require the storage class parameter to be present in upload requests, otherwise they will be rejected.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/28318

#### Screenshots
New Admin Console Input:
Storage Class Input

#### Release Note
```release-note
Added FileSettings.AmazonS3StorageClass and FileSettings.ExportAmazonS3StorageClass, both default to "" to preserve the current behaviour.
Administrators may configure this storage class to the storage class required by their S3 solution.
```

### CODE REVIEW

**Comment by NadavTasher** [comment]
@hanzei Can you help me with this?

**Comment by agnivade** [comment]
This seems like a reasonable addition. WDYT - @streamer45 @nab-77

**Comment by streamer45** [comment]
On a high level, it seems reasonable, but I am slightly concerned about the recommendations we could provide when a customer asks us about this setting and how it affects the application.

@NadavTasher Could you describe your use case in a bit more detail? Are you using something other than AWS S3 as remote storage? If so, how does the storage class come into play?

**Comment by NadavTasher** [comment]
Am using Ceph for object storage, and when the storage class provided does not match the storage class configured for the bucket, it just rejects the upload.

**Comment by NadavTasher** [comment]
Can anyone approve the workflows?

**Comment by NadavTasher** [comment]
I have fixed the failing test - should work now :)
Can you approve the execution once again?

**Comment by NadavTasher** [comment]
Is there anything that I can do to pass this Enterprise CI test?

**Comment by NadavTasher** [comment]
What's left to do for this PR?

**Comment by NadavTasher** [comment]
@streamer45 @agnivade 
What's left to do for this PR?

**Comment by streamer45** [comment]
> @streamer45 @agnivade What's left to do for this PR?

Review most likely. Let me assign a couple of people.

**Comment by agnivade** [comment]
I wanted to get some thoughts on @nab-77 first and see if this is okay to add.

**Comment by NadavTasher** [comment]
Are you able to provide any insight on the failed test?
Might as well fix them until @nab-77 has time to look at this

**Comment by NadavTasher** [comment]
@nab-77 - Will you be able to review my changes?
Until then - @agnivade / @streamer45 - Are you able to provide any insight on the failed test?

**Comment by agnivade** [comment]
The enterprise test failed because the PR is not up to date with latest master. But first, I'd like to get confirmation on whether we can proceed or not.

**Comment by agnivade** [comment]
1. "S3 Standard" is the default storage class, and the default config option should be that. I don't believe we need to change any QA scope for this.
2. Leaning towards an exposed system console option. Because the default setting does not change anything. It will keep using the default storage class, while allowing you to change it if you want to.

File: server/public/model/config.go
Comment (isacikgoz) [inline server/public/model/config.go:1769]: This should be an empty string right?

**Comment by NadavTasher** [comment]
Actually, the default (in the current codebase) is to completely skip sending the header, which would then lead to the default behaviour (that's why I kept it empty in the code I added).

Should I add code to pull this variable from the environment? (My guess was that the console & configuration options were sufficient)
Also, do I need to add any documentation? If so, what should I add and where?

File: server/public/model/config.go
Comment (NadavTasher) [reply server/public/model/config.go]: That is correct.
I have pushed a commit that fixes this.

**Comment by NadavTasher** [comment]
I have rebased on top of `master`.
Can anyone approve the CI workflows?

**Comment by NadavTasher** [comment]
I have rebased on top of master.
Can anyone approve the CI workflows?
@agnivade @streamer45

**Comment by NadavTasher** [comment]
Looks like the Enterprise tests pass now.
Should I add documentation in [https://github.com/mattermost/docs](https://github.com/mattermost/docs)?

**Comment by NadavTasher** [comment]
I have opened a documentation PR for this feature: https://github.com/mattermost/docs/pull/7467

**Comment by agnivade** [comment]
@NadavTasher - Please wait until we come to a consensus on this. Thanks.

**Comment by NadavTasher** [comment]
@amyblais Do you think this you can accept this PR?

**Comment by NadavTasher** [comment]
@nab-77 What's left here to do?

**Comment by nab-77** [comment]
Agree with you here @agnivade that the default must stay as is with System Console option available to make the change.

We can/should move ahead if and only if there is no need for any admins to change settings in the default case.

**Comment by agnivade** [comment]
Sounds like we have a consensus. Thanks @nab-77. We will proceed with the review.

**Comment by NadavTasher** [comment]
@agnivade Can you approve my workflow runs? I have rebased on top of master.

**Comment by NadavTasher** [comment]
Any updates? @agnivade @nab-77

**Comment by agnivade** [comment]
Apologies for the delay here. It's on my list. Please stay tuned.

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:1668]: We should update the code in `(s *FileSettings) isValid` to check that the storage class is one of `STANDARD | REDUCED_REDUNDANCY | STANDARD_IA | ONEZONE_IA | INTELLIGENT_TIERING | GLACIER | DEEP_ARCHIVE | OUTPOSTS | GLACIER_IR | SNOW | EXPRESS_ONEZONE`.  And do it for `ExportAmazonS3StorageClass` and `AmazonS3StorageClass` both.

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:1778]: There is no need to check for empty string. It's redundant.

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:1848]: Same as above.

**Review by agnivade — COMMENTED** [review]
Thanks for the PR @NadavTasher. Appreciate your patience. Left some comments.

**Comment by NadavTasher** [comment]
Thanks, @agnivade .

I have pushed commits that address your comments.
Notice that I have allowed empty strings in the config validation to preserve the current behaviour.

I have also updated the documentation PR (https://github.com/mattermost/docs/pull/7467) to account for these validations.

Do you think you could review my changes and approve the workflow run?

P.S. As a matter of opinion, I do not agree with this validation as these constants could change in the future, making Mattermost not future-proof. 
I'd prefer to have a non-validated value rather than a value validated against constants, this validation also adds unnecessary constants to the code.

**Comment by agnivade** [comment]
Thanks @NadavTasher. Just a quick FYI - we typically avoid force pushing, because it allows a reviewer to easily see what changed from the last code review. Just something to remember for next time :)

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:3984]: This can be re-written as

```go
if *s.AmazonS3StorageClass != "" && !slices.Contains([]string{..list of constants..}, s.AmazonS3StorageClass) {
  // return error
}
```

File: server/public/model/config.go
Comment (agnivade) [inline server/public/model/config.go:3988]: same as above.

**Review by agnivade — APPROVED** [review]
Thanks!

**Comment by NadavTasher** [comment]
Can you help me with the tests?

**Comment by agnivade** [comment]
Could you run `make i18n-extract`? Your added translation should get re-ordered in the json file.

**Comment by NadavTasher** [comment]
I have added this patch. Thanks!

**Comment by NadavTasher** [comment]
All tests have passed, should we merge?

**Comment by agnivade** [comment]
This needs to have another dev approval, then needs to go through QA. Appreciate your patience on this @NadavTasher.

**Review by isacikgoz — APPROVED** [review]
LGTM, thanks @NadavTasher

**Comment by agnivade** [comment]
@DHaussermann - Some QA steps:
1. Ensure that you see the new text setting in system console.
2. By default it should be empty.
3. Ensure that sanity testing on S3 works with the empty flag (You have to get credentials to an S3 bucket. If you don't have one, reach out to me).
4. Ensure that you see an error while trying to save an incorrect value for the setting.
5. Set other valid values, and do sanity tests.

**Comment by NadavTasher** [comment]
Any updates?

**Comment by NadavTasher** [comment]
Any updates on QA?

**Comment by DHaussermann** [comment]
Hi @NadavTasher I'll be looking at this shortly. I'm hopping it will fit into my day today.

**Comment by NadavTasher** [comment]
This can also be tested by spinning up a Ceph Object Gateway instance https://github.com/ceph/cn?tab=readme-ov-file#installation, but testing with Amazon seems like the best way to go.

**Comment by DHaussermann** [comment]
- Repeated testing with an AWS S3 bucket provided by Agniva
- Again tested leaving the new value blank as well as setting the value for storage class
- No issues setting/updating the config 
- In all cases for blank storage class or having the value set, no issues with using S3 as storage.

**Review by DHaussermann — APPROVED** [review]
Tested and passed
Approved based on testing above. 
No issues found.
LGTM!

Huge thanks @NadavTasher for implementing this 🎉

**Comment by NadavTasher** [comment]
Thank you for taking the time to review and test my code :)

**Pull request merged**
